File: apps/ios/ExpUI/Sources/Picker/StatusPicker.swift
import SwiftUI

// EXP-1029 contract — the status picker: the team's statuses (EXP-314) in
// display order, each by its glyph in its colour (`ResolvedStatus`, the
// status-icons rule). The issue header, the create-issue sheet and the bulk
// edit pick one.

public struct StatusPickerStatus: Identifiable, Hashable {
    public let id: String
    public let name: String
    /// Contract `issueStatusCategory`.
    public let category: String
    /// The resolved row colour (custom rows' hex, builtins' token).
    public let colorHex: String?
    /// The resolved glyph for the row (an `AppIcons` name).
    public let icon: String?
    /// EXP-1021: the resolved colour when it is a design TOKEN rather than a
    /// hex. A builtin row keeps the platform's status token — theme-aware,
    /// and no hex names it (EXP-314) — so it wins over `colorHex` when set.
    public let color: Color?

    public init(
        id: String,
        name: String,
        category: String,
        colorHex: String? = nil,
        icon: String? = nil,
        color: Color? = nil
    ) {
        self.id = id
        self.name = name
        self.category = category
        self.colorHex = colorHex
        self.icon = icon
        self.color = color
    }
}

public struct StatusPicker<Trigger: View>: View {
    public let statuses: [StatusPickerStatus]
    public let mode: PickerMode
    public let value: Set<String>
    public let onChange: (Set<String>) -> Void
    /// EXP-1021 — the surface controls every typed picker forwards verbatim
    /// (web's `PickerSurfaceProps`): a host that opens the picker from its own
    /// property row or `…` menu drives `open` and hides the trigger, and
    /// `onDismiss` fires once the sheet finished animating away (what a
    /// hand-off to a SECOND picker is promoted on).
    public let open: Binding<Bool>?
    public let hideTrigger: Bool
    public let onDismiss: (() -> Void)?
    private let trigger: () -> Trigger

    public init(
        statuses: [StatusPickerStatus],
        mode: PickerMode = .single,
        value: Set<String>,
        onChange: @escaping (Set<String>) -> Void,
        open: Binding<Bool>? = nil,
        hideTrigger: Bool = false,
        onDismiss: (() -> Void)? = nil,
        @ViewBuilder trigger: @escaping () -> Trigger
    ) {
        self.statuses = statuses
        self.mode = mode
        self.value = value
        self.onChange = onChange
        self.open = open
        self.hideTrigger = hideTrigger
        self.onDismiss = onDismiss
        self.trigger = trigger
    }

    nonisolated public static func items(_ statuses: [StatusPickerStatus]) -> [PickerItem<String>] {
        statuses.map { status in
            PickerItem(
                value: status.id,
                label: status.name,
                icon: status.icon,
                color: status.color ?? status.colorHex.flatMap { Color(hex: $0) },
                keywords: [status.name, status.category]
            )
        }
    }

    public var body: some View {
        GlassPicker(
            items: Self.items(statuses),
            mode: mode,
            value: value,
            onChange: onChange,
            emptyText: "No statuses",
            title: "Status",
            open: open,
            hideTrigger: hideTrigger,
            onDismiss: onDismiss,
            trigger: trigger
        )
    }
}

File: apps/android/app/src/main/java/com/exponential/app/ui/components/picker/StatusPicker.kt
package com.exponential.app.ui.components.picker

import androidx.compose.runtime.Composable
import androidx.compose.ui.graphics.Color
import com.exponential.app.ui.parseColor

/**
 * EXP-1029 contract, EXP-1021 implementation — the status picker: the team's statuses (EXP-314) in
 * display order, each by its glyph in its colour (the status-icons rule).
 * The issue header, the create-issue sheet and the bulk edit pick one.
 */
data class StatusPickerStatus(
    val id: String,
    val name: String,
    /** Contract `issueStatusCategory`. */
    val category: String,
    /** The resolved row colour (custom rows' hex, builtins' token). */
    val colorHex: String? = null,
    /**
     * EXP-1021: the RESOLVED glyph name + tint (`ResolvedIssueStatus.iconName`
     * and `resolvedStatusColor`, the web twin's `icon`). A builtin row's stored
     * hex is a near-neutral seed value that reads wrong, and a started row's
     * glyph is its position's pie clock — neither is derivable from the
     * category alone, so the caller hands both in.
     */
    val iconName: String? = null,
    val color: Color? = null,
)

fun statusPickerItems(statuses: List<StatusPickerStatus>): List<PickerItem<String>> =
    statuses.map { status ->
        PickerItem(
            value = status.id,
            label = status.name,
            icon = status.iconName?.let { com.exponential.app.ui.icons.ExpIcons.byName(it) },
            color = status.color ?: status.colorHex?.takeIf { it.isNotBlank() }?.let(::parseColor),
            keywords = listOf(status.name, status.category),
        )
    }

@Composable
fun StatusPicker(
    statuses: List<StatusPickerStatus>,
    value: Set<String>,
    onChange: (Set<String>) -> Unit,
    mode: PickerMode = PickerMode.Single,
    /** The sheet headline; the default names the picker. */
    title: String = "Status",
    /**
     * EXP-1021: the sheet CONTROLLED by the caller, for a picker that is a
     * state machine rather than a chip (the issue screens open theirs from a
     * properties sheet that has already closed).
     */
    open: Boolean? = null,
    onOpenChange: ((Boolean) -> Unit)? = null,
    trigger: @Composable (open: () -> Unit) -> Unit = {},
) {
    Picker(
        items = statusPickerItems(statuses),
        mode = mode,
        value = value,
        onChange = onChange,
        emptyText = "No statuses",
        title = title,
        open = open,
        onOpenChange = onOpenChange,
        trigger = trigger,
    )
}
